barretenberg/cpp/src/barretenberg/stdlib/special_public_inputs/special_public_inputs.hpp

```diff
@@ -98,9 +98,20 @@ class KernelIO {
      * @brief Set each IO component to be a public input of the underlying circuit.
      *
      */
-    void set_public()
+    void set_public(Builder* builder)
     {
-        pairing_inputs.set_public();
+        BB_ASSERT_EQ(builder, kernel_return_data.get_context());
+        BB_ASSERT_EQ(builder, app_return_data.get_context());
+        BB_ASSERT_EQ(builder, ecc_op_tables[0].get_context());
+        BB_ASSERT_EQ(builder, output_pg_accum_hash.get_context());
+
+        if (pairing_inputs.P0.get_context() == nullptr) {
```

Comment:
Are there legitimate cases where this happens? Would be nice if we could just disallow this and remove the `builder` arg from this method